### ISSUE TITLE
Avoid currentCompartment use for computing values of 

### ISSUE THREAD

**Issue by SergeStinckwich** [UMMISCO/kendrick#56]
Parameters defined in models  in a model are built in KENDRICK as a values or lexical closure like : 

```Smalltalk
model atParameter: #beta assignValue: 1.
```

```
model 
		atParameter: #mu
		assignValue: 
		[ :aModel| |c val| 
			c := aModel currentCompartment at: #species.
			c = #mosquito ifTrue: [ val := 12.17 ].
			c = #reservoir1 ifTrue: [ val := 0.05 ].
			c = #reservoir2 ifTrue: [ val := 0.05 ].
			val
		].
```

The problem is that for some models the parameters depends on the compartment (like the example above) or the time (seasonal forcing). At the moment, every time we execute a model, we have to do : ```model currentCompartment: aCompartment```, do update the context in order to be able to compute parameters values.

This is cumbersome and not a good idea to put this in the model ... How can we think about a better design for that ?

In fact, we have to select one value among several values depending on compartment values ?

**Comment by SergeStinckwich** [comment]
I think in fact that we don't need lexical closure.

We should be express a parameter depending on the value of the compartment like that: 

```Smalltalk
model 
   atParameter: #mu
   dependingOn: #species
   if: #mosquito value: 12.17
   if: #reservoir1 value: 0.05
   if: #reservoir2 value: 0.05.
```
But this is not correct Smalltalk at the moment. Any ideas ?

**Comment by SergeStinckwich** [comment]
For simple parameters that does not depends on compartment values, we should have : 

```Smalltalk
model setParameter: #beta as: 1.
```

What do you think about that ?

**Comment by SergeStinckwich** [comment]
How this is express in the DSL @npapoylias ?

**Comment by npapoylias** [comment]
**Currently** 

_mu_species: #(12.17 0.05 0.05);_

**In the future** (helvetia) with a proper [] indexing operator: 

_mu[species] = [ 12.17 , 0.05 , 0.05 ]_

This is the closest linguistic abstraction for the composition logic 
of the cartesian product (so in the api you can use setParameter:at:, 
setParameter:at:at: and so on):

**mu** becomes **mu[mosquito], mu[reservoir1], mu[reservoir2]**.
Short-handed in the DSL can be **mu[species] (as above) to assign
a collection**. 

If there was also a spatial dimension it would be 
**mu[mosquito][paris] for single assignment and mu[species][cities] 
for collection assignment** etc

You can find the example in the Mosquito project (in Scenario MosquitoParameters) 
of the zero-install (normally you can load all projects by opening the KendrickUI,
even if you did not zero install)

**Comment by SergeStinckwich** [comment]
DSL example is this (from Nick):

```Smalltalk
Scenario MosquitoParameters
on: 'Mosquito';
gamma: 52;
beta: 1;
mu_species: #(12.17 0.05 0.05);
rho_species: #(
#(0 0.02 0.02) 
#(0.02 0 0) 
#(0.02 0 0)
).```

What I don't understand is how you know that first element in mu_species array is #mosquito, second one #reservoir1 and last one #reservoir2. Based on their position in the array ? Very fragile !

**Comment by SergeStinckwich** [comment]
What is a bit strange, is that we use lexical closure when we express models in Smalltalk and not in DSL.

**Comment by npapoylias** [comment]
The Smalltalk API has lambdas cause the core model is not declarative and apparently not clearly decomposed from the simulation (ie see the #**current**Compartment msg-send in your code above, it smells like part of the simulation).

The DSL abstracts away from these as I explained above by using a declarative syntax (ie letting the user only worry about the model). You can do easily the same in the API as well, by encapsulating the closures in cleaner api-methods for the users. But the optimal solution is to have a: **Declarative Model and then a Simulation Visitor that are clearly decomposed** in the core.

PS There are alternatives for positional arguments in mu[species] = [ x , y , z]. That may seem less "fragile", such as mu[mosquito] = x etc or using dictionaries. On the other hand positional arguments are present in most major programming languages (for eg when you call a function or method the arguments are determined by position: f(x,y,z))

**Comment by SergeStinckwich** [comment]
Yes all my ranting about that started by #currenCompartment smells :-)
Simulation engines should not mix their states with the models.

I want to have a declarative API for the Smalltalk syntax also. This was my initial question. 
Regarding this, to you have comments ?

I'm not against positional arguments, but the problem is where is the positions are defined. In another Kendrick components ?

If we follow the positional arguments stance, we can have something like: 

```Smalltalk
model 
   setParameter: #mu
   dependingOn: #species
   as: #((#mosquito 12.17) (#reservoir1 0.05) (#reservoir2 as: 0.05)).
```

but does not scale for matrices ?

**Comment by SergeStinckwich** [comment]
and for single value: 
```Smalltalk
model setParameter: #beta as: 1.
```

**Comment by npapoylias** [comment]
Yes this is now a declarative api ;) The DSL takes it a bit further and automatically
determines the species components, to simplify the syntax.
